packages/zero-react/src/zero-provider.tsx

```diff
@@ -0,0 +1,77 @@
+import {
+  createContext,
+  useContext,
+  useEffect,
+  useState,
+  type ReactNode,
+} from 'react';
+import {Zero} from '../../zero-client/src/client/zero.ts';
+import type {Schema} from '../../zero-schema/src/builder/schema-builder.ts';
+import type {CustomMutatorDefs} from '../../zero-client/src/client/custom.ts';
+import type {ZeroOptions} from '../../zero-client/src/client/options.ts';
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+const ZeroContext = createContext<unknown | undefined>(undefined);
+
+export function useZero<
+  S extends Schema,
+  MD extends CustomMutatorDefs<S> | undefined = undefined,
+>(): Zero<S, MD> {
+  const zero = useContext(ZeroContext);
+  if (zero === undefined) {
+    throw new Error('useZero must be used within a ZeroProvider');
+  }
+  return zero as Zero<S, MD>;
+}
+
+export function createUseZero<
+  S extends Schema,
+  MD extends CustomMutatorDefs<S> | undefined = undefined,
+>() {
+  return () => useZero<S, MD>();
+}
+
+export type ZeroProviderProps<
+  S extends Schema,
+  MD extends CustomMutatorDefs<S> | undefined = undefined,
+> = (ZeroOptions<S, MD> | {zero: Zero<S, MD>}) & {
+  init?: (zero: Zero<S, MD>) => void;
+  children: ReactNode;
+};
+
+export function ZeroProvider<
+  S extends Schema,
+  MD extends CustomMutatorDefs<S> | undefined = undefined,
+>({children, init, ...props}: ZeroProviderProps<S, MD>) {
+  const [zero, setZero] = useState<Zero<S, MD> | undefined>(
+    'zero' in props ? props.zero : undefined,
```

Comment:
why not?

```suggestion
    props.zero,
```

Reply:
TypeScript complains about that. I don't understand why it does in this case. It has to do with the `|` type.